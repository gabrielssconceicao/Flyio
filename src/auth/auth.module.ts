import { Global, Module } from '@nestjs/common';
import { BcryptService } from './bcrypt.service';
import { HashingServiceProtocol } from './hashing/hashing.service';

@Global()
@Module({
  imports: [],
  providers: [
    {
      provide: HashingServiceProtocol,
      useClass: BcryptService,
    },
  ],
  exports: [HashingServiceProtocol],
})
export class AuthModule {}
