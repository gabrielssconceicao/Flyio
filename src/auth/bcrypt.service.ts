import * as bcrypt from 'bcryptjs';
import { HashingServiceProtocol } from './hashing/hashing.service';

export class BcryptService extends HashingServiceProtocol {
  async hash(password: string): Promise<string> {
    const salt = await bcrypt.genSalt();
    return bcrypt.hash(password, salt);
  }

  async compare(password: string, hash: string): Promise<boolean> {
    return bcrypt.compare(password, hash);
  }
}
