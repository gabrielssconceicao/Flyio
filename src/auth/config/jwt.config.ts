import { registerAs } from '@nestjs/config';

export default registerAs('jwt', () => {
  return {
    secret: process.env.JWT_SECRET,
    accessTokenExpiresIn: Number(
      process.env.JWT_ACCESS_TOKEN_EXPIRES_IN ?? '3600',
    ),
    refreshTokenExpiresIn: Number(
      process.env.JWT_REFRESH_TOKEN_EXPIRES_IN ?? '14400',
    ),
    audience: process.env.JWT_TOKEN_AUDIENCE,
    issuer: process.env.JWT_TOKEN_ISSUER,
  };
});
