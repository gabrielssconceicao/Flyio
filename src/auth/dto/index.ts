import { LoginDto } from './login.dto';
import { LoginResponseDto } from './login-response.dto';
import { RefreshTokenDto } from './refresh-token.dto';
import { TokenPayloadDto } from './token-payload.dto';

export { LoginDto, LoginResponseDto, RefreshTokenDto, TokenPayloadDto };
