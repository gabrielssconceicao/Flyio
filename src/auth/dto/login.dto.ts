import { IsNotEmpty, IsString } from 'class-validator';
import {
  dtoErrorMessages,
  DtoErrorType,
} from '../../common/utils/dto-error-messages';
export class LoginDto {
  @IsNotEmpty({
    message: dtoErrorMessages('Username or email', DtoErrorType.REQUIRED),
  })
  @IsString()
  usernameOrEmail: string;

  @IsNotEmpty({ message: dtoErrorMessages('Password', DtoErrorType.REQUIRED) })
  @IsString()
  password: string;
}
