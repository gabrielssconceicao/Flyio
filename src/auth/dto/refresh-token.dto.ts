import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsString } from 'class-validator';

export class RefreshTokenDto {
  @ApiProperty({
    example: 'krcio48hgGYIFYJfuyfJYAYu',
    description: 'Refresh token',
  })
  @IsString()
  @IsNotEmpty()
  refreshToken: string;
}
