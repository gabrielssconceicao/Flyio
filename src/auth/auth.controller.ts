import { Body, Controller, HttpCode, HttpStatus, Post } from '@nestjs/common';
import { ApiBody, ApiOperation, ApiResponse, PickType } from '@nestjs/swagger';
import { AuthService } from './auth.service';
import { LoginDto, LoginResponseDto, RefreshTokenDto } from './dto';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @ApiOperation({ summary: 'User Login ' })
  @ApiBody({
    type: LoginDto,
    description: 'User Login',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'Token',
    type: LoginResponseDto,
  })
  @ApiResponse({
    status: HttpStatus.UNAUTHORIZED,
    description: 'User or password invalid',
    schema: {
      example: {
        statusCode: HttpStatus.UNAUTHORIZED,
        message: 'User or password invalid',
        error: 'Unauthorized',
      },
    },
  })
  @ApiResponse({
    status: HttpStatus.FORBIDDEN,
    description: 'User not active',
    schema: {
      example: {
        statusCode: HttpStatus.FORBIDDEN,
        message: 'User not active',
        error: 'Forbidden',
      },
    },
  })
  @HttpCode(HttpStatus.CREATED)
  @Post('login')
  login(@Body() loginDto: LoginDto) {
    return this.authService.login(loginDto);
  }

  @ApiOperation({ summary: 'Refresh access token' })
  @ApiBody({
    description: 'Refresh token',
    type: RefreshTokenDto,
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'New access token',
    type: PickType(LoginResponseDto, ['accessToken']),
  })
  @ApiResponse({
    status: HttpStatus.FORBIDDEN,
    description: 'Token expired',
    schema: {
      example: {
        statusCode: HttpStatus.FORBIDDEN,
        message: 'Token expired',
        error: 'Forbidden',
      },
    },
  })
  @HttpCode(HttpStatus.CREATED)
  @Post('refresh-token')
  refreshToken(@Body() refreshTokendto: RefreshTokenDto) {
    return this.authService.refreshToken(refreshTokendto);
  }
}
