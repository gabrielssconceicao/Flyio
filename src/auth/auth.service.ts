import { Inject, Injectable, UnauthorizedException } from '@nestjs/common';
import { LoginDto } from './dto/login.dto';
import { PrismaService } from '../prisma/prisma.service';
import { HashingServiceProtocol } from './hashing/hashing.service';
import { JwtService } from '@nestjs/jwt';
import jwtConfig from './config/jwt.config';
import { ConfigType } from '@nestjs/config';
@Injectable()
export class AuthService {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly hashingService: HashingServiceProtocol,
    private readonly jwtService: JwtService,
    @Inject(jwtConfig.KEY)
    private readonly jwtConfiguration: ConfigType<typeof jwtConfig>,
  ) {}
  private throwUnauthorizedError() {
    throw new UnauthorizedException('User or password invalid');
  }
  async login(loginDto: LoginDto) {
    const user = await this.prismaService.user.findFirst({
      where: {
        OR: [{ username: loginDto.login }, { email: loginDto.login }],
      },
    });
    if (!user) {
      throw new this.throwUnauthorizedError();
    }
    const isPasswordValid = await this.hashingService.compare(
      loginDto.password,
      user.password,
    );
    if (!isPasswordValid) {
      this.throwUnauthorizedError();
    }
    const accessToken = await this.createToken(user);
    return { accessToken };
  }

  private async createToken(user): Promise<string> {
    const payload = { username: user.username, sub: user.id };
    return this.jwtService.signAsync(payload, {
      expiresIn: this.jwtConfiguration.jwtTtl,
      secret: this.jwtConfiguration.secret,
    });
  }
}
