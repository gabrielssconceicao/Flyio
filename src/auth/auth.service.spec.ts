import { Test, TestingModule } from '@nestjs/testing';
import { AuthService } from './auth.service';
import { userPrismaService } from '../prisma/mock/prisma.service.mock';
import { hashingServiceMock } from './mocks/hashing.service.mock';
import { jwtConfigurationMock } from './mocks/jwt.configuration.mock';
import { jwtServiceMock } from './mocks/jwt.service.mock';
import { HashingServiceProtocol } from './hashing/hashing.service';
import { JwtService } from '@nestjs/jwt';
import jwtConfig from './config/jwt.config';
import { ConfigType } from '@nestjs/config';
import { PrismaService } from '../prisma/prisma.service';
import { LoginDto } from './dto/login.dto';
import { generateLoginDtoMock } from './mocks/login.dto.mock';
import { ForbiddenException, UnauthorizedException } from '@nestjs/common';
describe('<AuthService />', () => {
  let service: AuthService;
  let prismaService: PrismaService;
  let hashingService: HashingServiceProtocol;
  let jwtService: JwtService;
  let jwtConfiguration: ConfigType<typeof jwtConfig>;
  let loginDto: LoginDto;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AuthService,
        {
          provide: PrismaService,
          useValue: userPrismaService,
        },
        {
          provide: HashingServiceProtocol,
          useValue: hashingServiceMock,
        },
        {
          provide: JwtService,
          useValue: jwtServiceMock,
        },
        {
          provide: jwtConfig.KEY,
          useValue: jwtConfigurationMock,
        },
      ],
    }).compile();

    service = module.get<AuthService>(AuthService);
    prismaService = module.get<PrismaService>(PrismaService);
    hashingService = module.get<HashingServiceProtocol>(HashingServiceProtocol);
    jwtService = module.get<JwtService>(JwtService);
    jwtConfiguration = module.get<ConfigType<typeof jwtConfig>>(jwtConfig.KEY);
    loginDto = generateLoginDtoMock();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
    expect(prismaService).toBeDefined();
    expect(hashingService).toBeDefined();
    expect(jwtService).toBeDefined();
    expect(jwtConfiguration).toBeDefined();
  });

  describe('<Login />', () => {
    it('should realize login', async () => {
      jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue({
        id: 'id-1',
        username: 'jDoe',
        password: 'hashed-password',
        active: true,
      } as any);
      jest.spyOn(hashingService, 'compare').mockResolvedValue(true);
      jest
        .spyOn(jwtService, 'signAsync')
        .mockResolvedValueOnce('access-token')
        .mockResolvedValueOnce('refresh-token');

      const result = await service.login(loginDto);
      expect(prismaService.user.findFirst).toHaveBeenCalled();
      expect(hashingService.compare).toHaveBeenCalledWith(
        loginDto.password,
        'hashed-password',
      );
      expect(jwtService.sign).not.toHaveBeenCalled();
      expect(jwtService.signAsync).toHaveBeenCalledTimes(2);
      expect(result).toEqual({
        accessToken: 'access-token',
        refreshToken: 'refresh-token',
      });
      expect(result).toMatchSnapshot();
    });

    it('should thow an UnauthorizedException if credentials are invalid', async () => {
      jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue(null);
      await expect(service.login(loginDto)).rejects.toThrow(
        UnauthorizedException,
      );
    });

    it('should thow an UnauthorizedException if password are invalid', async () => {
      jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue({} as any);
      jest.spyOn(hashingService, 'compare').mockResolvedValue(false);

      await expect(service.login(loginDto)).rejects.toThrow(
        UnauthorizedException,
      );
      expect(prismaService.user.findFirst).toHaveBeenCalled();
      expect(hashingService.compare).toHaveBeenCalled();
    });

    it('should thow an ForbiddenException if user is not active', async () => {
      jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue({
        id: 'id-1',
        password: 'hashed-password',
        active: false,
      } as any);
      jest.spyOn(hashingService, 'compare').mockResolvedValue(true);

      jest.spyOn(jwtService, 'sign').mockReturnValue('access-token');

      await expect(service.login(loginDto)).rejects.toThrow(ForbiddenException);
      expect(prismaService.user.findFirst).toHaveBeenCalled();
      expect(hashingService.compare).toHaveBeenCalled();
      expect(jwtService.sign).toHaveBeenCalledTimes(1);
      expect(jwtService.sign).not.toHaveBeenCalledWith(
        { id: 'id-1' },
        {
          expiresIn: jwtConfigurationMock.accessTokenExpiresIn,
        },
      );
    });
  });
});
