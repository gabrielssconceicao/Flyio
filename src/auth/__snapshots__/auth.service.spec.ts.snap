// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthService /> <Login /> should realize login 1`] = `
{
  "accessToken": "access-token",
  "refreshToken": "refresh-token",
}
`;
